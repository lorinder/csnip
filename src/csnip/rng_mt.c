#define HAVE_STDINT_H
#define HAVE_STDLIB_H
#define HAVE_UNISTD_H
#define HAVE_TIME_H

#define CSNIP_SHORT_NAMES
#include <csnip/rng.h>
#include <csnip/rng_mt.h>
#include <csnip/mem.h>

static void gen_seed(const csnip_rng* R,
					int nseed,
					const unsigned long int* seed)
{
	// XXX: We're assuming here that unsigned long int pointers
	// can be punned to uint32_t pointers.  Rethink this.
	rng_mt_seed(R->state,
			nseed * sizeof(unsigned long int) / sizeof(uint32_t),
			(uint32_t*)seed);
}

static unsigned long int gen_getnum(const csnip_rng* R)
{
	return rng_mt_getnum(R->state);
}

rng csnip_rng_mt_makerng(csnip_rng_mt_state* state)
{
	rng R = {
		.minval = 0,
		.maxval = 0xFFFFFFFFul,
		.state = state,
		.seed = gen_seed,
		.getnum = gen_getnum,
	};

	return R;
}



void csnip_rng_mt_seed(csnip_rng_mt_state* S,
			int nseed,
			const unsigned int* seed)
{
	S->state[0] = seed[0];
	uint32_t prev = S->state[0];
	for (int i = 1; i < CSNIP_RNG_MT_N; ++i) {
		S->state[i] = 1812433253 * (prev ^ (prev >> 30)) + i;
		prev = S->state[i];
	}
	S->next = S->state + CSNIP_RNG_MT_N;
}

#define MT_M		397

static uint32_t mixbits(uint32_t u, uint32_t v)
{
	return (u & 0x80000000) | (v & 0x7fffffff);
}

static uint32_t twist(uint32_t u, uint32_t v)
{
	const uint32_t matrix_A = 0x9908b0df;
	return (mixbits(u, v) >> 1) ^ (v & 1 ? matrix_A : 0);
}

static void
update_state(csnip_rng_mt_state* S)
{
	uint32_t* p = S->state;
	for (int count = CSNIP_RNG_MT_N - MT_M + 1; --count; ++p)
		*p = p[MT_M] ^ twist(p[0], p[1]);
	for (int count = MT_M; --count; ++p)
		*p = p[MT_M - CSNIP_RNG_MT_N] ^ twist(p[0], p[1]);
	*p = p[MT_M - CSNIP_RNG_MT_N] ^ twist(p[0], S->state[0]);
	S->next = S->state;
}

uint32_t csnip_rng_mt_getnum(csnip_rng_mt_state* S)
{
	if (S->next == S->state + CSNIP_RNG_MT_N)
		update_state(S);
	uint32_t r = *S->next++;

	/* Tempering */
	r ^= r >> 11;
	r ^= (r << 7) & 0x9d2c5680;
	r ^= (r << 15) & 0xefc60000;
	r ^= r >> 18;

	return r;
}
