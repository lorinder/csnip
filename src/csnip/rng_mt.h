#ifndef CSNIP_RNG_MT_H
#define CSNIP_RNG_MT_H

/** @file rng_mt.h
 *  @defgroup rng_mt	Mersenne Twister
 *  @{
 *
 *  Mersenne twister RNG implementation
 */

#include <stdint.h>

#include <csnip/rng.h>

#define CSNIP_RNG_MT_N		624

#ifdef __cplusplus
extern "C" {
#endif

typedef struct {
	uint32_t state[CSNIP_RNG_MT_N];
	uint32_t* next;
} csnip_rng_mt_state;

/** Initialize a generic RNG descriptor. */
csnip_rng csnip_rng_mt_makerng(csnip_rng_mt_state* state);

/** Seed the MT Random number generator. */
void csnip_rng_mt_seed(csnip_rng_mt_state* S,
			int nseed,
			const uint32_t* seed);

/** Produce the next output number. */
uint32_t csnip_rng_mt_getnum(csnip_rng_mt_state* S);

#ifdef __cplusplus
}
#endif

/** @} */

#endif /* CSNIP_RNG_MT_H */

#if defined(CSNIP_SHORT_NAMES) && !defined(CSNIP_RNG_MT_HAVE_SHORT_NAMES)
#define rng_mt_state		csnip_rng_mt_state
#define rng_mt_makerng		csnip_rng_mt_makerng
#define rng_mt_seed		csnip_rng_mt_seed
#define rng_mt_getnum		csnip_rng_mt_getnum
#define CSNIP_RNG_MT_HAVE_SHORT_NAMES
#endif /* CSNIP_SHORT_NAMES && !CSNIP_RNG_MT_HAVE_SHORT_NAMES */
