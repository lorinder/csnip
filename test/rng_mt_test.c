#include <stdio.h>

#define CSNIP_SHORT_NAMES
#include <csnip/rng_mt.h>

int main()
{
	rng_mt_state S;
	rng_mt_seed(&S, 1, (uint32_t[]){ 1234567 });

	/* Print the first few values */
	for (int i = 0; i < 10; ++i) {
		printf("%lu\n", (unsigned long)rng_mt_getnum(&S));
	}

	/* Advance and then print a later value */
	puts("...");
	for(int i = 0; i < 1000000000; ++i)
		rng_mt_getnum(&S);
	unsigned long v = rng_mt_getnum(&S);
	printf("%lu\n", v);

	/* Check the obtained value */
	int fail = 0;
	if (v != 3209008604u)
		fail = 1;

	return fail;
}
