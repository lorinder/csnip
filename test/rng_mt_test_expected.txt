1018032531
1997911679
32849524
1557424454
85170501
3999700962
1344722528
3988770941
427154528
262278029
...
3209008604
